File: leetcode-practice-typescript/src/dp/198-house-robber.ts
function rob(nums: number[]): number {
    if (nums.length === 1) {
        return nums[0];
    }
    // dp[i] = max(dp[i - 2] + i, dp[i - 1])
    const dp: number[] = [];
    dp[0] = nums[0];
    dp[1] = Math.max(nums[0], nums[1]);
    for (let i = 2; i < nums.length; i++) {
        dp[i] = Math.max(dp[i - 2] + nums[i], dp[i - 1]);
    }
    return dp[nums.length - 1];
}

console.log(rob([1, 2, 3, 1]));

File: leetcode-practice-python/src/dp/198_house_robber.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        # dp[i] = max(dp[i - 1], dp[i - 2] + i)
        if not nums:
            return 0
        n = len(nums)
        if n == 1:
            return nums[0]

        dp = [0] * (n + 1)
        dp[1] = nums[0]

        for i in range(2, n + 1):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i - 1])
        return dp[n]


if __name__ == "__main__":
    solution = Solution()
    print(solution.rob([1, 2, 3, 1]))
    print(solution.rob([2, 7, 9, 3, 1]))
